### ISSUE TITLE
[ROCm] test_conv_with_as_strided fp16 lowp noise floor on gfx90a

### ISSUE THREAD

**Issue by jeffdaily** [pytorch/pytorch#183581]
> **Status:** investigated, tolerance fix landed in #183573 (relaxes
> `(atol, rtol)` on the lowp comparison and removes the MI200 skip). This
> issue records the analysis behind that tolerance choice and the
> underlying MIOpen behavior. Filed as `has workaround` because the PR
> doesn't change MIOpen — it sets the test's expectations to match the
> realistic fp16 precision of the kernel.

## Summary

`test/inductor/test_torchinductor.py::GPUTests::test_conv_with_as_strided_cuda` has been skipped on MI200 (gfx90a) via `@skipIfRocmArch(MI200_ARCH)` since PR #170205, with no documented root cause beyond "Get MI200 CI green again." This report walks through a first-principles error analysis of the fp16 conv that the test exercises, derives a principled tolerance for the lowp comparison, and validates it empirically on gfx90a.

**Headline result:** the test failure is fundamental fp16 precision, not an MIOpen bug. The observed error floor (~`2^-9.5` ≈ 1.33e-3) is well predicted by tiled mixed-precision GEMM error analysis. Setting `atol = 2^-9 ≈ 2e-3` (= 2 × fp16 machine epsilon) and `rtol = 2e-3` admits the gfx90a result with 1.5× atol-only headroom and 3× combined-threshold headroom at unit magnitude.

---

## 1. The test and the observed failure

The test compiles a Module containing a 1×1 fp16 conv (256 → 384, no bias), pads the output with zeros, re-strides, and clones:

```python
self.kv = nn.Conv2d(256, 384, kernel_size=(1, 1), stride=(1, 1), bias=False)
# ...
convolution = self.kv(x)
constant_pad_nd = aten.constant_pad_nd(convolution, [2, 2, 2, 2], 0.0)
as_strided = aten.as_strided(constant_pad_nd, [8, 384, 2, 20, 12], ...)
as_strided_1 = aten.as_strided(as_strided, [8, 384, 2, 2, 12, 12], ...)
clone = aten.clone(as_strided_1, memory_format=torch.contiguous_format)
```

Inputs: `x = torch.randn(8, 256, 16, 16)`.

`self.common(..., check_lowp=True)` is `check_model_gpu` at `test/inductor/test_torchinductor.py:737`. It runs the model twice:

1. **fp32 pass:** eager fp32 vs Inductor-compiled fp32, with caller-supplied `(atol, rtol)` (default `None`, falling through to dtype defaults).
2. **fp16 (lowp) pass:** model is downcast to fp16, every fp32 input is downcast to fp16, then `check_model` is called with `reference_in_float=True`. The reference is computed in fp32 (the model + inputs are upcast inside `check_model`), the actual in fp16, and they're compared. Caller-supplied `rtol` is floored at 2e-3 (`rtol = max(2e-3, rtol)`) **only if it is not None**; `atol` is passed through unchanged.

When the MI200 skip is removed and the test runs on gfx90a, the lowp pass fails deterministically:

```
Mismatched elements: 7891 / 1769472 (0.4%)
Greatest absolute difference: 0.000244140625 at index (0, 54, 0, 0, 9, 2) (up to 1e-05 allowed)
Greatest relative difference: 15.2734375  at index (4, 58, 1, 0, 5, 6) (up to 0.001 allowed)
```

Re-running 5×: identical numbers every time. Deterministic drift, not a non-deterministic solver pick (the gfx950 failure that motivated PR #173368 was a different shape — a non-deterministic XDLOPS solver, fixed by `DeterministicGuard(True)`).

MIOpen logging confirms: gfx90a chooses `GemmFwd1x1_0_1` (rocBLAS-based hgemm) with or without `DeterministicGuard`. The same fix that works for gfx950 has no effect on gfx90a — the solver is already deterministic; the problem is the numerics of that solver.

Note the worst absolute difference is exactly `2^-12 = 0.000244140625`. The worst relative difference of 15.27 corresponds to an absolute difference of also `~2^-12` at an index where the fp32 reference is small (`0.000244 / 15.27 ≈ 1.6e-5`). This is the first hint that the error has a **near-constant absolute noise floor independent of output magnitude** — i.e. it is not pure fp16 quantization (which would scale with the value).

---

## 2. Forward error analysis of the fp16 dot product

The 1×1 conv reduces, at each spatial output, to a dot product of length `N = C_in = 256`:

  $y_{n,k,h,w} = \sum_{c=0}^{N-1} W_{k,c} \cdot x_{n,c,h,w}$

For the test:
- `x ∼ N(0, 1)` ⇒ `σ_x = 1`
- `W ∼ U(-1/√fan_in, 1/√fan_in) = U(-1/16, 1/16)` (PyTorch default `kaiming_uniform_(a=√5)` for `Conv2d`) ⇒ `σ_W² = (1/16)² / 3 = 1/(3·256)` ⇒ `σ_W ≈ 0.0361`
- Output std: `σ_y = σ_W · σ_x · √N = (1/√(3N)) · 1 · √N = 1/√3 ≈ 0.577`

fp16 constants:
- Unit roundoff `u₁₆ = 2⁻¹¹ ≈ 4.88e-4`
- Machine epsilon `ε₁₆ = 2 u₁₆ = 2⁻¹⁰ ≈ 9.77e-4`
- Smallest fp16 normal: `2⁻¹⁴ ≈ 6.10e-5`
- Subnormal spacing: `2⁻²⁴ ≈ 5.96e-8`

We compare the three plausible accumulation models.

### Model A — pure fp16 accumulation

Wilkinson's deterministic bound for an N-length dot product:

  $|\mathrm{fl}(w·x) - w·x| \le \gamma_N \sum_i |w_i x_i|$,  $\gamma_N = N u₁₆ / (1 - N u₁₆)$

For N=256, γ_N ≈ 0.143. Expected `Σ|wᵢxᵢ| ≈ N · E|W| · E|x| = 256 · (1/32) · √(2/π) ≈ 6.4`. Worst-case bound: `0.143 · 6.4 ≈ 0.9`, loose enough to be unhelpful.

The probabilistic estimate (assume rounding errors are independent, zero-mean, σ ≈ u₁₆ per op):

  $σ_\mathrm{err} ≈ u₁₆ \cdot \sqrt{N} \cdot σ_{wx} ≈ 4.88\text{e-4} \cdot 16 \cdot 0.0361 ≈ 2.8\text{e-4}$

This would predict a max-of-1.77M-Gaussian-sample worst case of roughly `5 · σ ≈ 1.4e-3`. Closer to what we see — but Model A is rare on real hardware. MFMA on gfx90a doesn't accumulate in fp16; it has an fp32 accumulator inside `v_mfma_f32_16x16x16_f16`.

### Model B — fp32 accumulation, single fp16 round on output

If accumulation is in fp32 throughout (no intermediate fp16 rounding) and only the final output is rounded to fp16:

- Accumulation error: `≤ u₃₂ · N · max|wᵢxᵢ| ≈ 6e-8 · 256 · 0.1 ≈ 1.5e-6` (negligible)
- Final fp16 quantization: `≤ u₁₆ · |y|`

This predicts error **proportional to |y|**. At |y| ~ 1, error ~ 5e-4. At |y| ~ 1.6e-5 (subnormal-fp16), error ~ 9e-9. But we observe ~`2⁻¹²` absolute error at small-magnitude indices, four orders of magnitude above what Model B predicts. So Model B is wrong.

### Model C — fp32 accumulator with periodic fp16 re-rounding

rocBLAS hgemm tiles K (input-channel dimension) into chunks of T ≈ 16. Each tile MFMAs in fp32, but the partial accumulator is **re-rounded to fp16** at tile boundaries before being added back in subsequent iterations (this is consistent with how mfma_f32_16x16x16_f16-based hgemm kernels accumulate across the K loop in practice).

There are `N/T = 16` re-roundings. Each introduces an error bounded by `u₁₆ · |partial sum|`. Partial sums at tile boundaries have std

  $σ_\mathrm{partial} \approx σ_W \cdot σ_x \cdot \sqrt{T} \approx 0.0361 \cdot 4 \approx 0.14$

Accumulated noise (independent zero-mean errors):

  $σ_\mathrm{accum} \approx u₁₆ \cdot \sqrt{N/T} \cdot σ_\mathrm{partial} \approx 4.88\text{e-4} \cdot 4 \cdot 0.14 \approx 2.7\text{e-4}$

This predicts:
- A **near-constant absolute noise floor** of magnitude ~`2 σ_accum ~ 5e-4` to `4 σ_accum ~ 1e-3`, independent of the final |y|.
- Worst case over 1.77M outputs (max-of-Gaussian inflation ~5×): `≈ 5 · 2.7e-4 = 1.35e-3 ≈ 2⁻⁹·⁵`.

Both features match what we observe.

---

## 3. Tolerance derivation

We want a principled `(atol, rtol)` for the lowp comparison. The `assertEqual` formula is

  $|y_\mathrm{actual} - y_\mathrm{ref}| \le \mathrm{atol} + \mathrm{rtol} \cdot |y_\mathrm{ref}|$

so `atol` is the floor at `|y_ref| → 0`, and `rtol · |y|` is the magnitude-proportional headroom.

**atol must dominate** Model C's noise floor over max-of-many samples. With the c ≈ 4–5 max-of-Gaussian factor for 1.77M samples (`√(2 ln 1.77M) ≈ 5.4`):

  $\mathrm{atol} \ge c \cdot u₁₆ \cdot σ_y = 5 \cdot 4.88\text{e-4} \cdot 0.577 \approx 1.4\text{e-3}$

Rounding to a clean fp16-natural value: `atol = 2⁻⁹ ≈ 1.95e-3 ≈ 2e-3`. Interpretation: **`atol = 2 × fp16 machine epsilon` = "2 fp16 ULPs at unit magnitude"**. This is a hardware-derived constant, not a magic number.

**rtol must dominate** the residual fp16 output-quantization error at the magnitudes where it kicks in (i.e. `|y| > atol / u₁₆`). For `|y| ≤ 1` (the bulk of this test's outputs), `rtol = 2e-3 = 4 × u₁₆` covers fp16 quantization with a 4× cushion. (Coincidentally `2e-3` is also the existing lowp-pass floor, so it is in any case the minimum rtol the test can have.)

Final derived tolerance:

  $\mathrm{atol} = 2 \varepsilon_{16} = 2^{-9} \approx 2\text{e-3}, \quad \mathrm{rtol} = 4 u_{16} = 2\text{e-3}$

At the worst-case |y|=1: combined threshold `2e-3 + 2e-3 = 4e-3`, which is **3× the empirically observed max error** (1.33e-3 — see §5). At |y| → 0 (subnormal range): threshold `2e-3`, which is **4.9× the observed max error in the subnormal bucket** (4.12e-4 — see §5).

---

## 4. Why `check_lowp=False` was rejected as the fix

`check_lowp=False` would disable the fp16 comparison entirely (a precedent exists: `test_split_cumsum`, `test_rnn_loop`, `test_batch_norm` all do this for known fp16 accuracy gaps). This option was rejected because the actual fp16 accuracy IS good enough; we just needed to ask for the right tolerance. Setting tolerances grounded in forward error analysis keeps the fp16 regression coverage — if a future kernel update on either gfx90a or gfx950 increases the error by, say, 4× from the current `2⁻⁹·⁵`, it would exceed `atol=2e-3` and we'd catch it.

---

## 5. Empirical validation on gfx90a

Two probes, run on a gfx90a:sramecc+:xnack- host.

### Probe 1 — raw conv noise floor

Build the Conv2d alone, run fp32 and fp16 versions, compute `err = |y_fp16 - y_fp32|`. With `torch.manual_seed(0)`:

```
max |err| = 1.326561e-03 (= 2^-9.56)
mean|err| = 1.576562e-04
std |err| = 1.246057e-04
|err_no_dg - err_dg|.max() = 0.0e+00     (DeterministicGuard is a no-op on gfx90a)
```

Per-magnitude bucket:

| `|y_fp32|` bucket            | count   | mean `|err|` | max `|err|` |
|-----------------------------:|--------:|-------------:|------------:|
| `[0, 2⁻¹⁴)`  (subnormal-fp16)|      53 |     1.307e-4 |    4.121e-4 |
| `[2⁻¹⁴, 2⁻¹⁰)`               |     984 |     1.332e-4 |    5.627e-4 |
| `[2⁻¹⁰, 2⁻⁶)`                |  16,037 |     1.270e-4 |    7.621e-4 |
| `[2⁻⁶, 2⁻⁴)`                 |  51,166 |     1.284e-4 |    6.820e-4 |
| `[2⁻⁴, 2⁻²)`                 | 196,257 |     1.304e-4 |    8.041e-4 |
| `[2⁻², 1)`                   | 457,095 |     1.572e-4 |    9.646e-4 |
| `[1, 4)`                     |  64,840 |     2.743e-4 |    1.327e-3 |

The **mean** error is essentially flat at ~`1.3e-4` (Model C signature). The **max** error grows weakly with magnitude — partly because the buckets contain different sample counts (more samples → larger max-of-Gaussian), partly because the `[1, 4)` bucket adds some fp16-quantization contribution (`u₁₆ · 4 ≈ 2e-3`, on the same order as the floor). `corr(|err|, |y|) = +0.285` confirms the picture: mostly flat noise floor with a weak magnitude-proportional component, exactly Model C.

### Probe 2 — Inductor adds no error

The full Module (conv + pad + as_strided + clone) was run three ways: fp32 eager, fp16 eager, fp16 compiled. With `torch.compile(model.half(), backend="inductor")`:

```
fp16 eager   vs fp32 eager:  max|err| = 1.327e-3,  mean = 1.094e-4
fp16 compiled vs fp32 eager: max|err| = 1.327e-3,  mean = 1.094e-4
Inductor delta (compiled - eager, fp16): max = 0.0, all elements equal
```

**Inductor adds zero error.** All of the noise is in MIOpen/rocBLAS. This rules out any Inductor lowering or layout-conversion bug as the source — if we want a tighter bound, we'd need a tighter rocBLAS hgemm, not a different inductor codegen.

### Mismatch counts under various tolerances (Probe 2 output, full pipeline)

| `(atol, rtol)`                   | mismatch count / 1,769,472 |
|---------------------------------:|---------------------------:|
| `(1e-5, 1e-3)` (current default) |                    204,805 |
| `(1e-5, 2e-3)` (lowp floor)      |                     65,112 |
| `(1e-3, 2e-3)` (derived, tight)  |                          0 |
| `(2e-3, 2e-3)` (derived, chosen) |                          0 |

Both `(1e-3, 2e-3)` and `(2e-3, 2e-3)` pass cleanly with seed 0, which is a *worse* seed than whatever the unit-test runner uses — the in-test failure was 0.4% mismatches, this probe shows 11.6% under the same thresholds.

---

## 6. Final test change

The fix in #183573 is:

```diff
--- a/test/inductor/test_torchinductor.py
+++ b/test/inductor/test_torchinductor.py
@@ -104,7 +104,6 @@ from torch.testing._internal.common_utils import (
     IS_MACOS,
     IS_X86,
     MACOS_VERSION,
-    MI200_ARCH,
     NAVI_ARCH,
     parametrize,
     serialTest,
@@ -12943,7 +12942,6 @@ def forward(self, arg0_1, ...):
             (torch.randn(32), torch.randn(32)),
         )

-    @skipIfRocmArch(MI200_ARCH)
     def test_conv_with_as_strided(self):
         class Model(nn.Module):
             def __init__(self) -> None:
@@ -12969,12 +12967,16 @@ def forward(self, arg0_1, ...):
                 )
                 return clone

-        # Avoid MIOpen's non-deterministic fp16 NHWC XDLOPS solver on gfx950.
+        # DeterministicGuard locks MIOpen's solver choice on gfx950. atol/rtol
+        # cover the fp16 hgemm noise floor on gfx90a.
         with DeterministicGuard(True, warn_only=True):
             self.common(
                 Model(),
                 (torch.randn(8, 256, 16, 16),),
                 check_lowp=not is_halide_backend(self.device),
+                atol=2e-3,
+                rtol=2e-3,
             )
```

8 verification runs (5× `test_torchinductor.py`, 3× `test_compile_subprocess.py`): all PASS, ~3.5s / ~13.3s per run respectively.

---

## 7. Why an MIOpen issue isn't being filed

The hgemm error on gfx90a is **consistent with the published behavior of `v_mfma_f32_16x16x16_f16`** when used inside a tiled K-loop with periodic fp16 re-rounding. The observed max error `2⁻⁹·⁵` is within `1σ` of the predicted `σ_accum · max-of-Gaussian-factor ≈ 1.5e-3`. Nothing about this points to a kernel bug — it is "fp16 GEMM, used in fp16, at fp16 precision." A different rocBLAS solver (e.g. one that uses a wider intermediate accumulator) could give tighter accuracy, but the *correct* fix for the unit test is to set tolerances that fit fp16, not to chase MIOpen for unrealistic precision.

If MIOpen behavior changes in the future and the error floor grows substantially beyond `2⁻⁹·⁵`, `atol=2e-3` will start failing and that will be the signal to investigate.

---

## 8. Open questions / future work

- **Cross-arch validation.** The same `(atol=2e-3, rtol=2e-3)` should also pass on gfx942 (MI300) and gfx950 (MI350) under `DeterministicGuard`. Only gfx90a hardware was available for this investigation; somebody with MI300/MI350 access should re-run the test on those archs to confirm. Expected: tighter actual error (different MFMA tile sizes), no failures.
- **Tolerance audit.** Other fp16 conv tests in `test_torchinductor.py` rely on the same lowp-pass defaults (`atol=1e-5, rtol=1e-3`). They pass today on gfx90a probably by luck of input statistics — smaller N, smaller `σ_y`, or both. A systematic audit using the same Model C framework would identify other latent gfx90a fragilities before they flake in CI.
- **MIOpen issue (deferred, not filed).** If gfx90a hgemm ever needs to match fp32 to fp16-quantization precision (instead of Model C precision), we'd need a `GemmFwd1x1_0_1` variant with a persistent fp32 accumulator across the entire K dimension. That has performance implications and is well outside the scope of a unit-test fix.

cc @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben
